crates/ruma-identifiers/src/matrix_uri.rs

```diff
@@ -0,0 +1,467 @@
+use std::{convert::TryFrom, fmt};
+
+use percent_encoding::{percent_decode_str, percent_encode, AsciiSet, CONTROLS};
+use ruma_identifiers_validation::{
+    error::{MatrixIdError, MatrixToError},
+    Error,
+};
+use url::Url;
+
+use crate::{EventId, RoomAliasId, RoomId, RoomOrAliasId, ServerName, UserId};
+
+const MATRIX_TO_BASE_URL: &str = "https://matrix.to/#/";
+// Controls + Space + reserved characters from RFC 3986. In practice only the
+// reserved characters will be encountered most likely, but better be safe.
+// https://datatracker.ietf.org/doc/html/rfc3986/#page-13
+const TO_ENCODE: &AsciiSet = &CONTROLS
+    .add(b':')
+    .add(b'/')
+    .add(b'?')
+    .add(b'#')
+    .add(b'[')
+    .add(b']')
+    .add(b'@')
+    .add(b'!')
+    .add(b'$')
+    .add(b'&')
+    .add(b'\'')
+    .add(b'(')
+    .add(b')')
+    .add(b'*')
+    .add(b'+')
+    .add(b',')
+    .add(b';')
+    .add(b'=');
+
+/// All Matrix Identifiers that can be represented as a Matrix URI.
+#[derive(Debug, PartialEq, Eq)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum MatrixId {
```

Comment:
If `MatrixId` is going to primarily be an enum representing things for the Matrix URI, then i'd like for it to be called such, so not `MatrixId`, but `MatrixUri`.

@jplatte thoughts?

Reply:
Ah right. But until there's a change in the spec this list will contain only identifiers, that's why we chose that.

Reply:
Per the chat; i figured out the current approach is a compromise.

I think this is fine as-is, we can always change it later if it doesnt fit.